packages/langium/src/lsp/type-hierarchy-provider.ts

```diff
@@ -0,0 +1,134 @@
+        const rootNode = document.parseResult.value;
+        const targetNode = findDeclarationNodeAtOffset(
+            rootNode.$cstNode,
+            document.textDocument.offsetAt(params.position),
+            this.grammarConfig.nameRegexp,
+        );
+        if (!targetNode) {
+            return undefined;
+        }
+
+        const declarationNode = this.references.findDeclarationNode(targetNode);
+        if (!declarationNode) {
+            return undefined;
+        }
+
+        return this.getTypeHierarchyItems(declarationNode.astNode, document);
+    }
+
+    protected getTypeHierarchyItems(targetNode: AstNode, document: LangiumDocument): TypeHierarchyItem[] | undefined {
+        const nameNode = this.nameProvider.getNameNode(targetNode);
+        const name = this.nameProvider.getName(targetNode);
+        if (!nameNode || !targetNode.$cstNode || name === undefined) {
+            return undefined;
+        }
+
+        return [
+            {
+                kind: SymbolKind.Class,
+                name,
+                range: targetNode.$cstNode.range,
+                selectionRange: nameNode.range,
+                uri: document.uri.toString(),
+                ...this.getTypeHierarchyItem(targetNode),
+            },
+        ];
+    }
+
+    protected getTypeHierarchyItem(_targetNode: AstNode): Partial<TypeHierarchyItem> | undefined {
+        return undefined;
+    }
+
+    supertypes(params: TypeHierarchySupertypesParams, _cancelToken?: CancellationToken): TypeHierarchyItem[] | undefined {
+        const document = this.documents.getOrCreateDocument(URI.parse(params.item.uri));
+        const rootNode = document.parseResult.value;
+        const targetNode = findDeclarationNodeAtOffset(
+            rootNode.$cstNode,
+            document.textDocument.offsetAt(params.item.range.start),
+            this.grammarConfig.nameRegexp,
+        );
+        if (!targetNode) {
+            return undefined;
+        }
+        return this.getSupertypes(targetNode.astNode);
+    }
+
+    /**
+     * Override this method to collect the supertypes for your language.
+     */
+    protected abstract getSupertypes(node: AstNode): TypeHierarchyItem[] | undefined;
+
+    subtypes(params: TypeHierarchySubtypesParams, _cancelToken?: CancellationToken): TypeHierarchyItem[] | undefined {
+        const document = this.documents.getOrCreateDocument(URI.parse(params.item.uri));
+        const rootNode = document.parseResult.value;
+        const targetNode = findDeclarationNodeAtOffset(
+            rootNode.$cstNode,
+            document.textDocument.offsetAt(params.item.range.start),
+            this.grammarConfig.nameRegexp,
+        );
+        if (!targetNode) {
+            return undefined;
+        }
+
+        const references = this.references.findReferences(targetNode.astNode, {
```

Comment:
It's true that type hierarchies are usually defined with references from the subtypes to the supertypes, but you could also define a language where it's the other way around. I suggest moving this call to the specific implementation instead of making assumptions whether you might need it.